### PR TITLE
Changes for OpenSAML3 migration

### LINKED ISSUE #7156: Migrate the version of OpenSAML from OpenSAML2 to OpenSAML3
### Description
Since OpenSAML2 reached its end of life 3 years back ([July 31, 2016](https://wiki.shibboleth.net/confluence/display/OpenSAML/Home)) for Java. Security fixes are no longer supported due to the situation we need to upgrade it to OpenSAML3 ASAP. The process involves conversion of the available repositories which are using OpenSAML2 to be converted to OpenSAML3. 

### The major changes from OpenSAML2 to OpenSAML3 includes:

- The structure of the OpenSAML dependencies has changed between version 2 and 3. OpenSAML 3.x is structured as a Maven multi-module project.
-  The Message Context has changed to become more modular. `MessageHandlers` has been added to process the messages using message context.
-  The syntax has been changed for some processes:
    - The syntax for using the `HTTPRedirectDeflateEncoder` has changed to a small degree, mainly due to the new design of the message context.
    - The syntax for sending SOAP messages has changed substantially.
- A lot of package names have changed from version 2 to 3.
- Some of the classes has been renamed for example:
   - The `EntityIdCriteria` class used for resolving credentials is renamed to `EntityIdCriterion`.
   - The `EncryptionParameters` class used for specifying encryption options is renamed to `DataEncryptionParameters`.
- Some methods are dropped out or either moved into another class. For example:
   - The `Configuration.validateNonSunJAXP` method has been removed as it is not longer necessary to validate this. 
   - `Configuration.validateJCEProviders` method has been replaced by the `init()` method of `JavaCryptoValidationInitialize`.
- Changes to Field names such as:
   - The `StatusCode.SUCCESS_URI` used in the Artifact Response has been renamed to `StatusCode.SUCCESS`.

### Resources for the migration
- https://wiki.shibboleth.net/confluence/display/OpenSAML/OSTwoDeveloperManual
- https://wiki.shibboleth.net/confluence/display/OS30/Home
- https://build.shibboleth.net/nexus/content/sites/site/java-opensaml/3.4.2/apidocs/
- https://nxp.evelopers.com/content/43dfe113-26c9-46c7-97b7-107c735a5073/?attach=false&language=&url=
- http://shibboleth.1660669.n2.nabble.com/Shibboleth-Developers-f1660781.html

### Suggested Labels
Type/Improvement, Priority/High, Component/SAML, Complexity/High

### Suggested Assignee
@deshankoswatte

### PR DESCRIPTION
**Pull request by deshankoswatte**
### Proposed changes in this pull request
Changes for OpenSAML3 migration

### When should this PR be merged
Not specified yet

### Prerequisites
- wso2/orbit#395 - Orbit for OpenSAML3
- wso2/carbon-identity-framework#2473 - Migration from OpenSAML 2.x to 3.x

Fixes: wso2/product-is#7156

### CODE REVIEW

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/DefaultSAML2SSOManager.java [hunk 1]
```diff
@@ -143,16 +143,11 @@ public static void doBootstrap() {
 
         /* Initializing the OpenSAML library */
         if (!bootStrapped) {
-            Thread thread = Thread.currentThread();
-            ClassLoader loader = thread.getContextClassLoader();
-            thread.setContextClassLoader(new DefaultSAML2SSOManager().getClass().getClassLoader());
```
Comment (deshankoswatte) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/DefaultSAML2SSOManager.java]: Check if this process should also persist with the changes

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/X509CredentialImpl.java [hunk 2]
```diff
@@ -271,7 +270,7 @@ public X509Certificate getEntityCertificate() {
     }
 
     @Override
-    public CredentialContextSet getCredentalContextSet() {
+    public CredentialContextSet getCredentialContextSet() {
         // TODO Auto-generated method stub
         return null;
     }
```
Comment (deshankoswatte) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/X509CredentialImpl.java]: Change the method to:

```
/***
 * get a credential context set
 * @return Method not supported so will return null
 */
public CredentialContextSet getCredentialContextSet() {
      return null;
}
```

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/util/SSOUtils.java [hunk 3]
```diff
@@ -184,10 +176,15 @@ public static void setSignature(RequestAbstractType request, String signatureAlg
         }
 
         org.apache.xml.security.Init.init();
+        Thread thread = Thread.currentThread();
```
Comment (deshankoswatte) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/util/SSOUtils.java:186]: Explain why the process should be done

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/pom.xml [hunk 4]
```diff
@@ -43,6 +43,16 @@
...
+            <exclusions>
+                <exclusion>
+                    <groupId>org.opensaml</groupId>
+                    <artifactId>opensaml1</artifactId>
+                </exclusion>
+                <exclusion>
+                    <groupId>org.opensaml</groupId>
+                    <artifactId>opensaml</artifactId>
+                </exclusion>
+            </exclusions>
```
Comment (deshankoswatte) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/pom.xml]: Change the order of exclusions

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/pom.xml [hunk 5]
```diff
@@ -59,6 +69,16 @@
...
```
Comment (deshankoswatte) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/pom.xml]: Change the order of exclusions

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/util/SSOUtils.java [hunk 6]
```diff
@@ -98,15 +98,8 @@ private SSOUtils() {
...
-            SecureRandomIdentifierGenerator generator = new SecureRandomIdentifierGenerator();
-            return generator.generateIdentifier();
-        } catch (NoSuchAlgorithmException e) {
-            log.error("Error while building Secure Random ID", e);
-            //TODO : throw exception and break the flow
-        }
-        return null;
+        RandomIdentifierGenerationStrategy generator = new RandomIdentifierGenerationStrategy();
+        return generator.generateIdentifier();
     }
```
Comment (deshankoswatte) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/util/SSOUtils.java]: Add space after method signature

File: features/org.wso2.carbon.identity.application.authenticator.samlsso.server.feature/pom.xml [hunk 7]
```diff
@@ -41,6 +41,16 @@
...
+            <exclusions>
+                <exclusion>
+                    <groupId>org.opensaml</groupId>
+                    <artifactId>opensaml1</artifactId>
+                </exclusion>
+                <exclusion>
+                    <groupId>org.opensaml</groupId>
+                    <artifactId>opensaml</artifactId>
+                </exclusion>
+            </exclusions>
```
Comment (deshankoswatte) [inline features/org.wso2.carbon.identity.application.authenticator.samlsso.server.feature/pom.xml]: Change order of exclusions

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/DefaultSAML2SSOManager.java (hunk 1 above)
Comment (deshankoswatte) [reply components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/DefaultSAML2SSOManager.java]: Kept as it is for now

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/util/SSOUtils.java (hunk 3 above)
Comment (deshankoswatte) [reply components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/util/SSOUtils.java:186]: Explained

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/util/SSOUtils.java (hunk 6 above)
Comment (deshankoswatte) [reply components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/util/SSOUtils.java]: Added

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/X509CredentialImpl.java (hunk 2 above)
Comment (deshankoswatte) [reply components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/X509CredentialImpl.java]: Fixed

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/pom.xml (hunk 5 above)
Comment (deshankoswatte) [reply components/org.wso2.carbon.identity.application.authenticator.samlsso/pom.xml]: Changed

File: features/org.wso2.carbon.identity.application.authenticator.samlsso.server.feature/pom.xml (hunk 7 above)
Comment (deshankoswatte) [reply features/org.wso2.carbon.identity.application.authenticator.samlsso.server.feature/pom.xml]: Changed

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/pom.xml (hunk 4 above)
Comment (deshankoswatte) [reply components/org.wso2.carbon.identity.application.authenticator.samlsso/pom.xml]: Changed

File: pom.xml [hunk 8]
```diff
@@ -369,9 +385,9 @@
         <axis2.wso2.version>1.6.1-wso2v37</axis2.wso2.version>
 
         <!-- Orbit versions-->
-        <opensaml.version>2.6.6</opensaml.version>
-        <opensaml2.wso2.version>2.6.6.wso2v3</opensaml2.wso2.version>
-        <opensaml2.wso2.osgi.version.range>[2.6.0,3.0.0)</opensaml2.wso2.osgi.version.range>
+        <opensaml.version>3.1.1</opensaml.version>
```
Comment (deshankoswatte) [inline pom.xml]: Change OpenSAML version to 3.3.1

Comment (deshankoswatte) [reply pom.xml]: Fixed

**Review by madurangasiriwardena — APPROVED** [review]

**Pull request merged**